_posts/2017-10-19-haskell-learning-group-2nd-meeting.md

```diff
@@ -0,0 +1,75 @@
+Something totally new for us was `$`, which has precedence `0`. It is an infix operator and it can be convenient for reducing amount of parentheses in expression. We tried to experiment with it and had a look on how it might be used. For instance, simple example:
+`(2^) (2 + 2)` _becomes_ `(2^) $ 2 + 2`  
+A bit more complex example with multiple dollar signs:
+`(2^) ((+2) ( 3*2))` _becomes_ `(2^) $ (+2) $ 3*2`
+
+## Strings
+
+Chapter 4 was all about strings and operations that can be performed on strings. It is important to understand that strings in Haskell are Lists of characters. There is a REPL command `:type` which gives type information on given statement. We tried it on string and we got `[Char]`. This means that "Hello" could be also represented as `['H', 'e', 'l', 'l', 'o']`. On the other hand, there is still `String` as a type alias, which can be used for instance when function take a string parameter as in example with sayHello: `sayHello :: String -> IO ()`.  
+By knowing that String is a List, we can perform operations like `++` for joining two Lists (or Strings if you wish). We were also playing a lot with such functions as `take` and `drop`. One of the exercises from the book was to write a function that swap the order of the words in the sentence. So the sentence "Curry is awesome" should be reversed to "awesome is Curry". One constraint we had - we are limited to use only functions `drop` and `take` to get the result. Of course it is not optimal for this case, but here is how the solution to that exercise looks like:
+
+```haskell
+rvrs :: String -> String
+rvrs x = ((drop 9 x) ++ (" " ++ (take 2 (drop 6 x)))) ++ (" " ++ (take 5 x))
+```
+
+
+
+## Exercises
+
+Some participants of a group who went a bit further with the book or even read the whole book before, decided to create a [repo with exercises][repo-exercises]. Project structure in a repository represents chapters in the book we are following. This is really practical, you know exactly where to stop or how you can practice things that you've just read about.
+The idea is that you have tests that are either pending or failing. Your task is to make them all green. For instance, here are the tests for [the third Chapter][chapter3-exercises].  
+You have to uncomment them or in some cases change `xit` to `it` to see them fail. Write whatever needed to make tests green.  
+If you are practicing TDD, it's not a right time to follow suggestions "write as less code as possible just to make test pass". So don't cheat and if expectation is that function `add` with arguments 1 and 2 should equal 3, don't just return 3 in the body of the Function ;).  
```

Comment:
Ok I guess I  will  drop the whole paragraph, Its not that important. The idea was to not only concentrate to make tests green.  So if there is an expectation that add 1 2 should  return 3, don't end up  with `let add = 3`, because there will be no more further tests that will prove that function implemented for other cases